File: 7scanner/scanner.py
"""
Disciplina: Compiladores - PUC-SP
Scanner léxico para tokenizar texto em português.

Autores: 
Isabela Nunes 
Pedro Huck
Sthefany Viveiros
"""

import re
import sys

def tokenizar(texto):
    """
    Quebra o texto em tokens usando expressões regulares.
    Tokens reconhecidos:
    - Palavras com acentos
    - Palavras com hífen
    - Números
    - Pontuação

    Parâmetro:
        texto (str): string contendo todo o conteúdo do arquivo de entrada.

    Retorno:
        list: lista contendo todos os tokens encontrados no texto.
    """

    regex = r"""
        \.\.\. |                    # reticências
        \d+(?:[.,]\d+)? |          # números
        [A-Za-zÀ-ÖØ-öø-ÿ]+(?:-[A-Za-zÀ-ÖØ-öø-ÿ]+)* |  # palavras
        [.,!?;:"()\[\]—-]          # pontuação
    """

    tokens = re.findall(regex, texto, re.VERBOSE)
    return tokens



def main():
    """
    Função principal:
    - Recebe arquivo de entrada por argumento
    - Gera tokens
    - Salva no output.txt
    """

    #Verifica se o usuário informou o arquivo de entrada
    #sys.argv contém os argumentos passados na execução do programa
    if len(sys.argv) < 2:
        print("Uso: python scanner.py <arquivo.txt>")
        return

    
    arquivo_entrada = sys.argv[1]    #Nome do arquivo de entrada recebido pela linha de comando
    arquivo_saida = "output.txt"    #Nome do arquivo onde os tokens serão salvos

    try:
        #Abre o arquivo de entrada em modo leitura
        #encoding UTF-8 garante suporte a caracteres acentuados
        with open(arquivo_entrada, "r", encoding="utf-8") as f:
            texto = f.read()
    except FileNotFoundError:
        #Tratamento caso o arquivo não exista
        print("Erro: arquivo de entrada não encontrado.")
        return

    tokens = tokenizar(texto)    #Chama a função que realiza a tokenização do texto

    print("Primeiros 50 tokens:")
    print(tokens[:50])    #Mostra no terminal apenas os primeiros 50 tokens

    with open(arquivo_saida, "w", encoding="utf-8") as out:    #Abre o arquivo de saída para escrita
        for token in tokens:
            out.write(token + "\n")    #Cada token será salvo em uma linha

    #informações finais para o usuário
    print(f"\nTotal de tokens: {len(tokens)}")
    print("Tokens salvos em output.txt")


if __name__ == "__main__":
    main()

File: 7scanner/Scanner.java
import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.util.regex.*;

/**
 * Disciplina: Compiladores - PUC-SP
 * Scanner léxico para tokenizar textos em português.
 *
 * Entrada:
 *     Arquivo .txt (UTF-8)
 *
 * Saída:
 *     Arquivo output.txt (já existente), contendo um token por linha
 */


/**
 * Método responsável pela tokenização do texto.
 *
 * Utiliza expressões regulares (regex) para identificar padrões
 * linguísticos e dividir o texto em tokens.
 *
 * @param texto Conteúdo completo do arquivo de entrada
 * @return Lista de tokens identificados no texto
 */
public class Scanner {
    public static List<String> tokenizar(String texto) {
        //Lista onde os tokens encontrados serão armazenados
        List<String> tokens = new ArrayList<>();

        String regex =
                "\\.\\.\\." +                 // reticências
                "|\\d+(?:[.,]\\d+)?" +       // números
                "|[A-Za-zÀ-ÖØ-öø-ÿ]+(?:-[A-Za-zÀ-ÖØ-öø-ÿ]+)*" + // palavras
                "|[.,!?;:\"()\\[\\]—-]";     // pontuação

        Pattern pattern = Pattern.compile(regex);    //compila a expressão regular em um padrão
        Matcher matcher = pattern.matcher(texto);    //cria um matcher para buscar padrões dentro do texto

        while (matcher.find()) {    //Percorre todo o texto procurando correspondências com a regex
            tokens.add(matcher.group());
        }

        return tokens;    //Retorna a lista de tokens encontrados
    }



    /**
     * Método principal do programa.
     *
     * Responsabilidades:
     * - Ler o arquivo de entrada informado via argumento
     * - Executar o scanner léxico
     * - Exibir alguns tokens no terminal
     * - Salvar todos os tokens no arquivo output.txt
     *
     * @param args Argumentos da linha de comando
     */
    public static void main(String[] args) {

        if (args.length < 1) {    //Verifica se o usuário forneceu o arquivo de entrada
            System.out.println("Uso: java Scanner <arquivo.txt>");
            return;
        }

        String arquivoEntrada = args[0];
        String arquivoSaida = "output.txt"; // já existe

        try {
            // leitura do arquivo
            String texto = Files.readString(Paths.get(arquivoEntrada));

            // tokenização
            List<String> tokens = tokenizar(texto);

            //exibe os primeiros 50 tokens no terminal
            System.out.println("Primeiros 50 tokens:");
            for (int i = 0; i < Math.min(50, tokens.size()); i++) {
                System.out.print(tokens.get(i) + " ");
            }

            // escrita no output.txt
            BufferedWriter writer = new BufferedWriter(new FileWriter(arquivoSaida));

            for (String token : tokens) {
                writer.write(token);
                writer.newLine();
            }

            writer.close();

            System.out.println("\n\nTotal de tokens: " + tokens.size());
            System.out.println("Tokens salvos em output.txt");

        } catch (IOException e) {    //Tratamento de erro
            System.out.println("Erro ao ler ou escrever arquivo: " + e.getMessage());
        }
    }
}
